pkg/applicationserver/redis/registry.go

```diff
@@ -174,18 +175,21 @@ func (r *DeviceRegistry) Set(ctx context.Context, ids ttnpb.EndDeviceIdentifiers
 					return errInvalidIdentifiers
 				}
 			} else {
-				if err := ttnpb.ProhibitFields(sets,
-					"ids.application_ids",
-					"ids.dev_eui",
-					"ids.device_id",
-					"ids.join_eui",
-				); err != nil {
-					return errInvalidFieldmask.WithCause(err)
-				}
-
 				if err := cmd.ScanProto(updated); err != nil {
 					return err
 				}
+				if ttnpb.HasAnyField(sets, "ids.application_ids") && !pb.ApplicationIdentifiers.Equal(updated.ApplicationIdentifiers) {
+					return errReadOnlyField.WithAttributes("field", "ids.application_ids")
+				}
+				if ttnpb.HasAnyField(sets, "ids.device_id") && pb.DeviceID != updated.DeviceID {
+					return errReadOnlyField.WithAttributes("field", "ids.device_id")
+				}
+				if ttnpb.HasAnyField(sets, "ids.join_eui") && (pb.JoinEUI == nil || !updated.JoinEUI.Equal(*pb.JoinEUI)) {
```

Comment:
`JoinEUI` can be `nil`, if it's not stored. `pb.JoinEUI == nil` condition here is incorrect